File: src/lessons/recursion/cons/PlusOneEntity.java
package lessons.recursion.cons;

import static plm.core.ValueSerializer.*;

import lessons.recursion.cons.universe.ConsEntity;
import lessons.recursion.cons.universe.RecList;

public class PlusOneEntity extends ConsEntity {

  public void run()
  {
    int count = getTestCount();
    for (int i = 0; i < count; i++) {
      Object[] param = (Object[])deserialize(getTest(i));
      setTestResult(i, serialize(plusOne(RecList.fromArray((int[]) param[0]))));
    }
  }

  /* BEGIN TEMPLATE */
  RecList plusOne(RecList seq)
  {
    /* BEGIN SOLUTION */
    if (seq == null)
      return null;
    return cons(seq.head + 1, plusOne(seq.tail));
    /* END SOLUTION */
  }
  /* END TEMPLATE */
}

File: src/lessons/recursion/cons/PlusOneEntity.py
def run():
    count = getTestCount()
    for i in range(count):
        param = deserialize(getTest(i))
        setTestResult(i, serialize(plusOne(toRecListIfArray(param[0]))))

# BEGIN TEMPLATE
def plusOne(list):
# BEGIN SOLUTION
  if list == None:
    return None
  return cons(list.head+1, plusOne(list.tail))
# END SOLUTION
# END TEMPLATE
